tensorflow_probability/python/distributions/johnson_su.py

```diff
@@ -0,0 +1,335 @@
+      dtype = dtype_util.common_dtype([gamma, delta, loc, scale], tf.float32)
+      self._gamma = tensor_util.convert_nonref_to_tensor(
+          gamma, name='gamma', dtype=dtype)
+      self._delta = tensor_util.convert_nonref_to_tensor(
+          delta, name='delta', dtype=dtype)
+      self._loc = tensor_util.convert_nonref_to_tensor(
+          loc, name='loc', dtype=dtype)
+      self._scale = tensor_util.convert_nonref_to_tensor(
+          scale, name='scale', dtype=dtype)
+      dtype_util.assert_same_float_dtype((self._gamma, self._delta,
+                                          self._loc, self._scale))
+      super(JohnsonSU, self).__init__(
+          dtype=dtype,
+          reparameterization_type=reparameterization.FULLY_REPARAMETERIZED,
+          validate_args=validate_args,
+          allow_nan_stats=allow_nan_stats,
+          parameters=parameters,
+          name=name)
+
+  @staticmethod
+  def _param_shapes(sample_shape):
+    return dict(
+        zip(('gamma', 'delta', 'loc', 'scale'),
+            ([tf.convert_to_tensor(sample_shape, dtype=tf.int32)] * 4)))
+
+  @classmethod
+  def _params_event_ndims(cls):
+    return dict(gamma=0, delta=0, loc=0, scale=0)
+
+  @property
+  def gamma(self):
+    """Gamma shape parameters in these Johnson's SU distribution(s)."""
+    return self._gamma
+
+  @property
+  def delta(self):
+    """Delta shape parameters in these Johnson's SU distribution(s)."""
+    return self._delta
+
+  @property
+  def loc(self):
+    """Locations of these Johnson's SU distribution(s)."""
+    return self._loc
+
+  @property
+  def scale(self):
+    """Scaling factors of these Johnson's SU distribution(s)."""
+    return self._scale
+
+  def _batch_shape_tensor(self, gamma=None, delta=None, loc=None, scale=None):
+    return functools.reduce(
+        prefer_static.broadcast_shape,
+        (prefer_static.shape(self.gamma if gamma is None else gamma),
+         prefer_static.shape(self.delta if delta is None else delta),
+         prefer_static.shape(self.loc if loc is None else loc),
+         prefer_static.shape(self.scale if scale is None else scale)))
+
+  def _batch_shape(self):
+    return functools.reduce(
+        tf.broadcast_static_shape,
+        (self.gamma.shape, self.delta.shape, self.loc.shape, self.scale.shape))
+
+  def _event_shape_tensor(self):
+    return tf.constant([], dtype=tf.int32)
+
+  def _event_shape(self):
+    return tf.TensorShape([])
+
+  def _sample_n(self, n, seed=None):
+    gamma = tf.convert_to_tensor(self.gamma)
+    delta = tf.convert_to_tensor(self.delta)
+    loc = tf.convert_to_tensor(self.loc)
+    scale = tf.convert_to_tensor(self.scale)
+    batch_shape = self._batch_shape_tensor(gamma=gamma, delta=delta,
+                                           loc=loc, scale=scale)
+    shape = tf.concat([[n], batch_shape], axis=0)
+    seed = SeedStream(seed, 'johnson_su')
+
+    u = tf.random.uniform(shape, minval=0., maxval=1.,
```

Comment:
I'd recommend instead using `tf.random.normal` and feeding that in to the `tf.sinh`. The sampling algorithm for `tf.random.normal` will be more efficient than using `ndtri`.

Reply:
Switched to TransformedDistribution as per your suggestion so this code was removed. Looks much better now, thanks :+1: